File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxDepth(root) {

        if(root == null)
        {
            return 0;
        }

        const queue = [];
        queue.push(root);
        let levels = 0;

        while(queue.length > 0)
        {
            levels++;
            const size = queue.length;
            
            for(let i = 0; i < size; i++)
            {
            const poppedNode = queue.shift();
            if(poppedNode.left != null) queue.push(poppedNode.left);
            if(poppedNode.right != null) queue.push(poppedNode.right);

            }


        }

        return levels;
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    private TreeNode helper(TreeNode root, int k, int[] counter) {
        if (root == null) {
            return null;
        }

        TreeNode node = helper(root.left, k, counter); // Traverse the left subtree
        if (node != null) return node;

        counter[0]++; // Increment the counter when visiting a node
        if (counter[0] == k) return root;

        return helper(root.right, k, counter); // Traverse the right subtree
    }

    public int kthSmallest(TreeNode root, int k) {
        int[] counter = new int[1]; // Array to hold the counter reference
        TreeNode node = helper(root, k, counter);
        return node != null ? node.val : -1; 
    }
}
